### PR TITLE
[Old - don't merge] Cloud / Amazon error handling improvement

### PR DESCRIPTION
**Pull request by mikedlr**
##### SUMMARY

This is a proposed new function for handling error messages from AWS in a standardised way.  The aim would be that all cloud/amazon modules should use it.  This is now part of https://github.com/ansible/ansible/pull/25780

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module_utils

##### ANSIBLE VERSION

```
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/mikedlr/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mikedlr/development/ansible/ansible-add-modules/lib/ansible
  executable location = /home/mikedlr/development/ansible/ansible-add-modules/bin/ansible
  python version = 2.7.11 (default, Jul  8 2016, 19:45:00) [GCC 5.3.1 20160406 (Red Hat 5.3.1-6)]
```

##### ADDITIONAL INFORMATION

This has no user visible changes at present.  Later error messages should be improved.

### CODE REVIEW

**Comment by willthames** [comment]
So much +1. 

Probably worth checking if any modules can drop their import of `camel_dict_to_snake_dict` after using this (if you haven't already)

**Comment by gundalow** [comment]
If people like this then https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md will need updating

**Comment by gundalow** [comment]
+1 to concept

**Comment by s-hertel** [comment]
I like the idea of this a lot. Not sure how I feel about this being in basic.py. It would be a little more complicated but make more sense to have this in module_utils/ec2.py I think, in which case we would either need to pass the function the module to use fail_json or else have a two step process where we call the function to get the contents to then call fail_json. What are your thoughts on that? There are some minor things to fix as well like not using str and `fail_json(msg=message, traceback=last_traceback)` should be `fail_json(msg=message, exception=last_traceback)`. Discussed with @abadger.

File: lib/ansible/module_utils/basic.py [hunk 1]
```diff
@@ -2027,6 +2028,37 @@ def fail_json(self, **kwargs):
...
+        else:
+            message = except_msg
+
+        try:
+            response=exception.response
+        except AttributeError:
+            response=None
+
+        if response is None:
+            self.fail_json(msg=message, traceback=last_traceback)
```
Comment (s-hertel) [inline lib/ansible/module_utils/basic.py]: should be exception=last_traceback

File: lib/ansible/module_utils/basic.py [hunk 2]
```diff
@@ -2027,6 +2028,37 @@ def fail_json(self, **kwargs):
...
+        else:
+            self.fail_json(msg=message, traceback=last_traceback,
+                       **camel_dict_to_snake_dict(response))
```
Comment (s-hertel) [inline lib/ansible/module_utils/basic.py]: you could line that up with msg

File: lib/ansible/module_utils/basic.py (hunk 1 above)
Comment (mikedlr) [reply lib/ansible/module_utils/basic.py]: very funny.. I'm testing for the wrong thing so I write completely the wrong thing

File: lib/ansible/module_utils/basic.py (hunk 2 above)
Comment (mikedlr) [reply lib/ansible/module_utils/basic.py]: Also done a complete flake8 cleanup of my new code and the tests in response to this.  However basic.py is a complete mess and I left most of that alone for now.

**Comment by mikedlr** [comment]
@s-hertel as far as having this in basic.py goes:

Probably the real solution to moving everything to the ec2 module would be to make the a special AnsibleAWSModule class which is a subclass of (or very similar to a subclass of) AnsibleModule.  Then all the functions could be moved there.  I'd propose doing that by creating a new file `lib/ansible/module_utils/aws.py` and doing a big refactoring to use it.  This will be much easier to achieve if all the use of error reporting is regular at that time.  So to move this forward I propose:

- accept this patch to basic.py as is
- update the AWS module guidelines, but say that it only applies to modules which are merged to the main repo.  Outside modules should not use that fail_json_aws yet.
- when someone volunteers, create the new file - `aws.py` - containing the `AnsibleAWSModule` class and put this + all important ec2 connection functions there (renaming where appropriate), 
- Migrate all aws modules to use the `AnsibleAWSModule`
- finally delete fail_json_aws from basic.py where it will no longer be needed

This will avoid making the use of fail_json_aws difficult and complex at any time.

**Comment by abadger** [comment]
<nod> a subclass makes sense. However, we need to be conscious of (but not block on) the fact that we're hoping to create a new API which includes an AnsibleModule that's much lighter weight than what's in basic.AnsibleModule right now.

Here's an idea (please evaluate, though, it's my first thought this morning).  Perhaps using composition instead of inheritance would be appropriate here?  I know that the Docker modules use composition.   I'm thinking something like this:

``` python
class AnsibleAWSModule(object):
    def __init__(self, [all sorts of params]):
        self._module = AnsibleModule([appropriate_params])

    @property
    def params(self):
        return self._module.params

    def exit_json(self, *args, **kwargs):
        return self._module.exit_json(*args, **kwargs)

    def fail_json(self, *args, **kwargs):
        return self._module.exit_json(*args, **kwargs)

    def fail_json_aws(self, exception, msg=None):
        # Your implementation [...]
```

The idea is to control the API that is exposed in the AnsibleAWSModule by not giving access to all of the functionality of AnsibleModule.  Later, when AnsibleModule is split up into pieces, it should be reasonable to change this into inheritance and only have the AnsibleAWSModule API gain methods, not lose them (that is, as long as we don't choose to expose anything from AnsibleModule that subsequently gets moved...)

**Comment by mikedlr** [comment]
This is now effectively part of https://github.com/ansible/ansible/pull/25780 which is likely to be merged soon.

**Comment by mikedlr** [comment]
This is now merged via https://github.com/ansible/ansible/pull/25780 so I'm closing this pull request.

**Pull request closed without merging**